### PR TITLE
Suppress "Failed to detach context" warning from OpenTelemetry

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/13914?quickstart=1)

#### Install mlflow from this PR

```
# Use `%sh` to run this command on Databricks
OPTIONS=$(if pip freeze | grep -q 'mlflow @ git+https://github.com/mlflow/mlflow.git'; then echo '--force-reinstall --no-deps'; fi)
pip install $OPTIONS git+https://github.com/mlflow/mlflow.git@refs/pull/13914/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13914
```

</p>
</details>

### What changes are proposed in this pull request?

When OpenTelemetry span is created in one thread (or async task) and then finished in another thread, it raises the following warning:

```
ValueError: <Token var=<ContextVar name='current_context' default={} at 0x7ff7c4a1be20> at 0x7ff79095f540> was created in a different Context
```

OpenTelemetry manages active span via [ContextVar](https://docs.python.org/3/library/contextvars.html), which is isolated between threads or async tasks. This warning claims the span is created with one context, but then finished in different context.

While this is sometimes problematic, it does not necessarily has impact to the tracing. For example, LangChain does many tricky handling for context e.g., run some code within a cloned context, even if it is not multi-thread or async execution. In such case, this warning is shown but the trace is created perfectly, which will confuse users.

Since it is not realistic for us to workaround all of those different context hacks in 3P libraries, this PR just address the issue by demoting the warning to debug level, so it won't be surfaced to users at least.

In case if this issue actually break tracing, it should be captured by current test suites (we also have a [validation](https://github.com/mlflow/mlflow/blob/b527225ff53e82c9d7a0c425f079fc5382d2bb58/tests/conftest.py#L74-L91) for 'leaked' span context at every test case).

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

**Before**:
Screenshot 2024-11-29 at 14 08 20

**After**:
Screenshot 2024-11-29 at 14 08 13

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/langchain/langchain_tracer.py [hunk 1]
```diff
@@ -31,9 +32,37 @@
 VS_INDEX_DOC_URL_COL = "doc_uri"
 
 
-# Langchain config Keys
-CONFIGURABLE_KEY = "configurable"
```
Comment (B-Step62) [inline mlflow/langchain/langchain_tracer.py:35]: I forgot to address comment here: https://github.com/mlflow/mlflow/pull/13828#discussion_r1853379665

File: mlflow/tracing/utils/token.py [hunk 2]
```diff
@@ -17,3 +18,32 @@ class SpanWithToken:
...
     token: Optional[contextvars.Token] = None
+
+
+class LogDemotionFilter(logging.Filter):
+    def __init__(self, module: str, message: str):
+        super().__init__()
+        self.module = module
+        self.message = message
+
+    def filter(self, record):
```
Comment (harupy) [inline mlflow/tracing/utils/token.py:29]: Is it possible to type hint record?

Comment (harupy) [reply mlflow/tracing/utils/token.py:29]:
```suggestion
    def filter(self, record: ???) -> bool:
```

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/tracing/utils/token.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/tracing/utils/token.py:29]: Yea it should be `logging. LogRecord`

**Review by TomeHirata — APPROVED** [review]
LGTM

**Pull request merged**